### ISSUE TITLE
API not creating nor updating variation images

### ISSUE THREAD

**Issue by seanlane** [woocommerce/woocommerce#13729]
## EXPLANATION OF THE ISSUE

Tested using Wordpress 4.7.3 and Woocommerce 2.6.14, on a new WP installation at Dreamhost as well as a locally running installation, with both the Twenty Seventeen and Twenty Twelve themes.

I would like to use the API to create several variations of a product, including setting the image that corresponds to the color of that variation (the product in mind is a t-shirt, which can come in one of several colors). Using the Woocommerce Python API module, I wrote a script to do that. I verified the request data, and everything there seems to be in order, so I believe this to be an issue with Woocommerce Core as opposed to the Python module.

I generate the data to be sent to the API, which includes at least one variation which has one image, and the product can either have images or have none at all. After sending the request to the API, the response is as follows:

1. If there were images for the main product, the variant image is set to the primary image of the product.
2. If the product did not have an image, then the variant image is set to the placeholder image, like the product itself.

I would expect that the variant image should be set to the image specified in the data sent in the original request.

Furthermore, when editing the product within the dashboard, the variant shows that no image was set, not even to the main product image. This contradicts the response received from the API which indicated that the variant image was set to the primary image of the product, even though that is incorrect behavior as well. This can be verified by looking at the `postmeta` table of the MySQL database for the Wordpress installation and seeing that no `thumbnail_id` is set for the variant. However, if you manually set the variant image within the dashboard, this record will be set within the MySQL database, which I believe is the correct behavior.

## STEPS TO REPRODUCE THE ISSUE

1. Create a new Wordpress installation
2. Attempt to add a product via the API that includes at least one variant which has a corresponding image
3. View the response of the API that states the variant's image is set to the image of the product itself
4. View in the dashboard that the variant image is never actually set

## SYSTEM STATUS REPORT

```
### WordPress Environment ###

Home URL: http://localhost:8888/wordpress
Site URL: http://localhost:8888/wordpress
WC Version: 2.6.14
Log Directory Writable: ✔
WP Version: 4.7.3
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US

### Server Environment ###

Server Info: Apache
PHP Version: 7.1.0
PHP Post Max Size: 32 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.51.0
OpenSSL/1.0.2j

SUHOSIN Installed: –
MySQL Version: 5.6.34
Max Upload Size: 32 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 2.6.14
: 
woocommerce_sessions: ✔
woocommerce_api_keys: ✔
woocommerce_attribute_taxonomies: ✔
woocommerce_downloadable_product_permissions: ✔
woocommerce_order_items: ✔
woocommerce_order_itemmeta: ✔
woocommerce_tax_rates: ✔
woocommerce_tax_rate_locations: ✔
woocommerce_shipping_zones: ✔
woocommerce_shipping_zone_locations: ✔
woocommerce_shipping_zone_methods: ✔
woocommerce_payment_tokens: ✔
woocommerce_payment_tokenmeta: ✔
MaxMind GeoIP Database: ❌ The MaxMind GeoIP Database does not exist - Geolocation will not function. You can download and install it manually from http://dev.maxmind.com/geoip/legacy/geolite/ to the path: . Scroll down to \"Downloads\" and download the \"Binary / gzip\" file next to \"GeoLite Country\"

### Active Plugins (1) ###

WooCommerce: by WooThemes – 2.6.14

### Settings ###

Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2

### API ###

API Enabled: ✔

### WC Pages ###

Shop Base: #4 - /shop/
Cart: #5 - /cart/
Checkout: #6 - /checkout/
My Account: #7 - /my-account/

### Taxonomies ###

Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)


### Theme ###

Name: Twenty Twelve
Version: 2.2
Author URL: https://wordpress.org/
Child Theme: ❌ – If you're modifying WooCommerce on a parent theme you didn't build personally
then we recommend using a child theme. See: How to create a child theme

WooCommerce Support: ✔

### Templates ###

Overrides: –
```

## Python Script to reproduce results

I have written the following Python 3 script that has only a dependency on the Woocommerce Python module. To install the module with pip, use the following command:

```
$ pip3 install --user woocommerce
```

Then run the following script. It will add a dummy product with a few pictures, and you can verify my findings either through the standard output messages or a debugger. Note that you will need to also enable the Woocommerce API, create API keys, and then insert the keys and the appropriate URL at the indicated locations in the beginning lines of the script.

``` python
from woocommerce import API

wcapi = API(
    url='INSERT_URL_HERE',
    consumer_key='INSERT_CONSUMER_KEY_HERE',
    consumer_secret='INSERT_CONSUMER_SECRET_HERE',
    wp_api=True,
    version="wc/v1",
    query_string_auth=True,
    timeout=300
)


# Add some categories

# Create a parent category "Cat" if doesn't exists
# Then get id
categories = wcapi.get("products/categories").json()
cat_categories = [x for x in categories if x['name'] == 'Cat']

parent_id = None
if not cat_categories:
    response = wcapi.post("products/categories", {'name': 'Cat'})
    assert response.ok, "Response not ok"
    parent_id = response.json()['id']
else:
    assert len(cat_categories) == 1, "More than one cat category for some reason"
    parent_id = cat_categories[0]['id']

for cat in ['Fluffy', 'Big', 'Kitten']:
    data = {
        "name": cat,
        "parent": parent_id
    }

    print(wcapi.post("products/categories", data).json())

# Add some attributes

print('Get attributes')
attributes = {x['name']: x for x in wcapi.get("products/attributes").json()}

print('Create attributes if needed')
for attr in ['Color', 'Size']:
    if attr not in attributes:
        r = wcapi.post('products/attributes', {'name': attr})
        assert r.ok, 'Response not ok for {} attribute'.format(attr)
        print(r.json())
        attributes[attr] = r.json()

# Get the color names
color_names = ['Orange', 'Brown', 'Black', 'White']

print('Add Color terms if needed')
color_attr_url = "products/attributes/{}/terms".format(attributes['Color']['id'])
color_terms = {x['name']: x for x in wcapi.get(color_attr_url).json()}

for color in color_names:
    if color in color_terms:
        continue

    data = {'name': color}
    print(wcapi.post(color_attr_url, data).json())

print('Add Size terms if needed')
size_attr_url = "products/attributes/{}/terms".format(attributes['Size']['id'])
size_terms = {x['name']: x for x in wcapi.get(size_attr_url).json()}

for val in ['Small', 'Medium', 'Large', 'X-Large', 'XX-Large']:
    if val in size_terms:
        continue

    data = {'name': val}
    print(wcapi.post(size_attr_url, data).json())

# Add some cats

dimensions = {
    "Small": {
        "length": "27",
        "width": "18"
    },
    "Medium": {
        "length": "28",
        "width": "20"
    },
    "Large": {
        "length": "29",
        "width": "22"
    },
    "X-Large": {
        "length": "30",
        "width": "24"
    },
    "XX-Large": {
        "length": "31",
        "width": "26"
    }
}

# Get attributes
attributes = {x['name']: x for x in wcapi.get("products/attributes").json()}

# Get attribute terms
attribute_data = []
colors = []
sizes = []
color_attribute_id = size_attribute_id = print_color_attr_id = None
for key in ['Color', 'Size']:
    terms = wcapi.get("products/attributes/{}/terms".format(attributes[key]['id'])).json()
    terms = [x['name'] for x in terms]
    attribute_data.append({
        "id": attributes[key]['id'],
        "variation": True,
        "options": terms
    })
    if key == 'Color':
        colors = terms
        color_attribute_id = attributes[key]['id']
    elif key == 'Size':
        sizes = terms
        size_attribute_id = attributes[key]['id']

# Get the cat categories from WP
categories = wcapi.get("products/categories?per_page=75").json()
categories = {x['name']: x for x in categories}
parent_id = categories['Cat']['id']

cats = {}
for key in categories.keys():
    if categories[key]['parent'] == parent_id:
        cats[key] = categories[key]

base_image_url = 'https://placekitten.com/g/'

resolutions = {
    'Orange': 525,
    'Brown': 550,
    'Black': 575,
    'White': 600
}

# Create variations
variation_data = []
for color in colors:
    for size in sizes:
        variation_data.append({
            "regular_price": '20.00',
            "visible": True,
            "dimensions": dimensions[size],
            "images": [
                {
                    "src": '{}/{}/{}'.format(base_image_url, resolutions[color], resolutions[color]),
                    "position": 0,
                    "name": size + ' - ' + color
                }
            ],
            "attributes": [
                {
                    "id": color_attribute_id,
                    "option": color
                },
                {
                    "id": size_attribute_id,
                    "option": size
                }
            ]
        })

# Create image data
image_data = []
for i in range(len(colors)):
    image_data.append(
        {
            "src": '{}/{}/{}'.format(base_image_url, resolutions[colors[i]], resolutions[colors[i]]),
            "position": i,
            "name": size + ' - ' + colors[i]
        })

data = {
    "name": " Cat",
    "type": "variable",
    "description": "This is a cat",
    "short_description": "Cat",
    "categories": [{"id": cats['Fluffy']['id']}],
    "attributes": attribute_data,
    "images": [image_data[0], image_data[1], image_data[2]],
    "variations": [variation_data[1]]
}

r = wcapi.post("products", data)
response = r.json()

req = r.request

print('Main Product Image Id: ', response['images'][0]['id'])
print('Variant Image Id: ', response['variations'][0]['image'][0]['id'])
print('These are the same image for some reason.')
print('Try updating the variant image:')

var_id = response['variations'][0]['id']
prod_id = response['id']
image_id = response['images'][2]['id']

var_data = {
    "variations": [
        {
            "id": var_id,
            "images": [image_data[2]]
        }
    ]
}

r = wcapi.put("products/" + str(prod_id), var_data)
print('Main Product Image Id: ', r.json()['images'][0]['id'])
print('Variant Image Id: ', r.json()['variations'][0]['image'][0]['id'])
print('Note that they\'re still the same')

# Uncomment to cleanup and allow for quick reiteration
# print(wcapi.delete('products/{}?force=true'.format(prod_id)).json())
```

**Comment by mikejolley** [comment]
@claudiosanches can you confirm?

**Comment by claudiosanches** [comment]
I followed the steps to reproduce, but no luck, working just fine, at least on 3.0.0 (master).

https://cloudup.com/cO9ZVM9A7zp

**Comment by seanlane** [comment]
Thanks for the response, I believe I found the cause of my issue. When creating the variation data, I was referring to the `image` property as `images`. For example:

```python
var_data = {
    "variations": [
        {
            "id": var_id,
            "images": [image_data[2]]
        }
    ]
}
``` 

instead of 

```python
var_data = {
    "variations": [
        {
            "id": var_id,
            "image": [image_data[2]]
        }
    ]
}
```

This caused the behavior originally described in Woocommerce version 2.6.14, and in version 3.0.0 RC1, the response will show that the variation's image is set to the placeholder image. Thank you for your help, and sorry for the errant issue!

**Comment by claudiosanches** [comment]
@seanlane just check the docs:

https://woocommerce.github.io/woocommerce-rest-api-docs/#variation-properties
![screenshot from 2017-03-23 17-28-14](https://cloud.githubusercontent.com/assets/1264099/24268887/1b4d25ea-0fee-11e7-8280-2e7455daa948.png)

And have an example of creating a variation too:
https://woocommerce.github.io/woocommerce-rest-api-docs/?python#create-a-product
![screenshot from 2017-03-23 17-30-00](https://cloud.githubusercontent.com/assets/1264099/24268988/5f51a7de-0fee-11e7-87e2-218a9778dc01.png)

**Comment by eduaero1** [comment]
Hello!
I am facing the exact same problem. I am not able to upload the variations pictures

I have tried modifying the product variations by:

var_data = {
                "variations": [
                    {
                        "id": 13898,
                        "image": 'https://website/wp-content/uploads/eva2.jpg'
                    }
                ]
            }
print(wcapi.put("products/13897", var_data).json())
AND 
print(wcapi.put("products/13897/variations/13898", var_data).json())

Furthermore I have tried by calling naming 'image' and 'images', without being successful...

I have tried to upload when I upload the variations data but it did not work neither... 

I would really appreciate any help as I have been working on this for the last week and I am totally stuck...

Thank you very much beforehand :)

**Comment by mikejolley** [comment]
`image` takes an array of data. Again, check the docs - there are example requests in the sidebar.

**Comment by eduaero1** [comment]
Thank you very much :)
I think that by now I know the documentation by heart :)
I upload all my products through the API REST. I'll add the steps that I am taking because I think that it should work (at least it works for everything less for the images variations...)
This is what I upload for creating the product
`data={
    "description": "<style>\n.t1 {<br />\n    border-collapse: collapse;<br />\n    width: auto;<br />\n}<br />\n.t2 {<br />\n    border-collapse: collapse;<br />\n    width: 100%;<br />\n}<br />\n.th1 {<br />\n    border: 1px solid #dddddd;<br />\n    text-align: left;<br />\n    padding: 8px;<br />\n}<br />\n</style>\n\nCamiseta de hombre con cuello redondo ribeteado de algod\u00f3n con doble pespunte. Doble costura en mangas, cintura y cuello.\nPeso: camiseta blanca 160 g y camiseta de color 165 g.\nTejido 100% algod\u00f3n.\nLos colores pueden variar de la pantalla del ordenador al color real. \n\nDise\u00f1o realizado por <a href=\"https://www.instagram.com/daxus_oficial/\">@daxus_oficial</a>\n<b>Lavado:</b> para tener m\u00e1s informaci\u00f3n de c\u00f3mo cuidar tus nuevos productos puedes leer <a href=\"http://xn--visteydisea-beb.com/cuidar-tus-productos/\">este post</a> para sacar el m\u00e1ximo provecho de tus personalizaciones.\n\n<strong>Dimensiones:\u00a0</strong>\n<table class=\"t2\">\n<tbody>\n<tr>\n<th>\n<table class=\"t1\">\n<tbody>\n<tr>\n<th class=\"th1\"></th>\n<th class=\"th1\">S</th>\n<th class=\"th1\">M</th>\n<th class=\"th1\">L</th>\n<th class=\"th1\">XL</th>\n<th class=\"th1\">XXL</th>\n<th class=\"th1\">3XL</th>\n</tr>\n<tr>\n<th class=\"th1\">Ancho (cm)</th>\n<th class=\"th1\">48,5</th>\n<th class=\"th1\">53</th>\n<th class=\"th1\">56</th>\n<th class=\"th1\">61</th>\n<th class=\"th1\">66</th>\n<th class=\"th1\">71</th>\n</tr>\n<tr>\n<th class=\"th1\">Largo (cm)</th>\n<th class=\"th1\">69,5</th>\n<th class=\"th1\">72</th>\n<th class=\"th1\">74,5</th>\n<th class=\"th1\">77</th>\n<th class=\"th1\">78,5</th>\n<th class=\"th1\">80</th>\n</tr>\n</tbody>\n</table>\n</th>\n</tr>\n</tbody>\n</table>\nLas medidas del largo son tomadas desde el punto m\u00e1s alto de la camiseta, se puede dar un m\u00e1rgen de error de 2,5 cm en ambas medidas.",
    "meta_data": [
        {
            "key": "_yoast_wpseo_primary_product_cat",
            "value": "",
            "id": 435340
        },
        {
            "key": "page_product_layout",
            "value": "inherit",
            "id": 435341
        },
        {
            "key": "page_product_youtube",
            "value": "",
            "id": 435342
        },
        {
            "key": "product_full_screen_description_meta_box_check",
            "value": "off",
            "id": 435343
        },
        {
            "key": "_wpb_vc_js_status",
            "value": "false",
            "id": 435344
        },
        {
            "key": "_vc_post_settings",
            "value": "{vc_grid_id: []}",
            "id": 435345
        },
        {
            "key": "_yoast_wpseo_focuskw_text_input",
            "value": "camiseta",
            "id": 435350
        },
        {
            "key": "_yoast_wpseo_focuskw",
            "value": "camiseta",
            "id": 435351
        },
        {
            "key": "_yoast_wpseo_metadesc",
            "value": "Camiseta de hombre con cuello redondo ribeteado de algod\u00c3\u00b3n con doble pespunte. Doble costura en mangas, cintura y cuello. Peso: camiseta blanca 160 g y camiseta de color 165 g. Tejido 100% algod\u00c3\u00b3n.",
            "id": 435352
        },
        {
            "key": "_yoast_wpseo_content_score",
            "value": 90,
            "id": 435353
        }
    ],
    "slug": "prueba-eva-hombre-rugby",
    "categories": [
        {
            "id": "152"
        },
        {
            "id": "246"
        },
        {
            "id": "237"
        }
    ],
    "attributes": [
        {
            "position": 1,
            "visible": "true",
            "variation": "True",
            "id": 3,
            "options": [
                "S"
            ]
        },
        {
            "position": 2,
            "visible": "true",
            "variation": "True",
            "id": 5,
            "options": [
                "Azul"
            ]
        }
    ],
    "images": [
        {
            "src": "https://xn--visteydisea-beb.com/wp-content/uploads/Camiseta-mujer-verde-botella-Jane.jpg",
            "position": 0
        }
    ],
    "name": "Camiseta hombre Rugby",
    "type": "variable",
    "short_description": ""
}

r_out = wcapi.post("products", data).json()
new_prod_id = r_out['id']
`

Now for the variation I upload the following code:
`variation={'attributes': [{'id': 3, 'option': 'S'}, {'id': 5, 'option': 'Azul'}], 'image': [{'src': 'https://xn--visteydisea-beb.com/wp-content/uploads/Camiseta-mujer-verde-botella-Jane.jpg', 'position': 0}], 'regular_price': '13,99'}
print(wcapi.post("products/%d/variations" % new_prod_id, variation).json())
`

Furthermore, having a look of what you have upload I have also tried to upload the following code:

`var_data = {
                "variations": [
                    {
                        "id": 13898,
                        "image": 'https://xn--visteydisea-beb.com/wp-content/uploads/Camiseta-mujer-verde-botella-Jane.jpg'
                    }
                ]
            }
            print(wcapi.put("products/13897", var_data).json())`

AND 
`print(wcapi.put("products/13897/variations/13898", var_data).json())`

However none if these works...
Any help it would be super welcome... this is driving me crazy... :(

Thank you very much beforehand guys :)

**Comment by eduaero1** [comment]
It does neither work by adding this...
However I think that position and scr should be necessary...

Product ID=13943
Variation ID= 13944
`var_data = {"variations": [{ "id": 13944,"image": ["https://xn--visteydisea-beb.com/wp-content/uploads/Camiseta-mujer-verde-botella-Jane.jpg"]}]}`

`print(wcapi.put("products/13943", var_data).json())`

`print(wcapi.put("products/13943/variations/13944", var_data).json())`

If someone could tell me a code that would work I would be super grateful, this is all I need :)

Cheers :)

**Comment by mikejolley** [comment]
It's hard to follow what you're doing exactly, your JSON isn't formatted well here, but have you considered creating a variable product manually and looking at the response via GET? What you send will look very similar to that!

**Comment by eduaero1** [comment]
Thank you very much Mike!!
I was totally stuck and I was not thining properly. In case that it is helpful for anyone this is what I have done :) (using Python)
prodID=952 #product ID
varID=953 #variation ID
`print(wcapi.get("products/952/variations" ).json())`

I got 
`{ ....
'image': {'src': 'Blue.jpg'}}`

So I upload the image by doing
`var_data2={'image': {'src': 'Red.jpg'}}`
`print(wcapi.put("products/952/variations/953",var_data2).json())`

Thank you very much sincerely. It was super helpful!!
